File: src/scenes/Splash.cpp
#include "Splash.hpp"
#include "Play.hpp"
#include "CppInvaders.hpp"
#include "pico.h"
#include <string>

#define TIME_TYPEWRITE 0.05f
#define TIME_WAITING 1.5f

static std::string LINES[] = {
    "PLAY",
    "SPACE INVADERS",
    "*SCORE ADVANCES TABLE*",
    "=? MYSTERY",
    "=30 POINTS",
    "=20 POINTS",
    "=10 POINTS",
    ""};

static Pico_Abs_Pos OFFS[] = {
    {96, 64}, {56,  88}, {24, 120},  {80, 136},
    {80, 152}, {80, 168}, {80, 184}, { 0,  0}};

SplashScene::SplashScene() {
    state = WAITING1;
    timer = 0;
    li = 0;
    ci = 0;
}

SplashScene::~SplashScene() {}

bool SplashScene::typewrite_next_char() {
    if (timer < TIME_TYPEWRITE) {
        return false;
    }

    if (ci++ == LINES[li].length()) {
        li++;
        ci = 0;
    }
    timer -= TIME_TYPEWRITE;
    return true;
}

void SplashScene::process_event(const Pico_Event &event) {
    if (event.type != PICO_EVENT_KEY_DOWN) {
        return;
    }

    if (state == WAITING_KEYPRESS) {
        CppInvaders &game = CppInvaders::get();
        game.scene = new PlayScene;
        game.score = 0;
        delete this;
    } else {
        state = WAITING_KEYPRESS;
        li = 7;
        ci = 0;
    }
}

void SplashScene::update(float delta) {
    timer += delta;

    switch (state) {
    case WAITING1:
        if (timer >= TIME_WAITING) {
            state = TYPEWRITING1;
            timer -= TIME_WAITING;
        }
        break;
    case TYPEWRITING1:
        if (typewrite_next_char() && li == 2) {
            state = WAITING2;
        }
        break;
    case WAITING2:
        if (timer >= TIME_WAITING) {
            state = TYPEWRITING2;
            timer -= TIME_WAITING;
        }
        break;
    case TYPEWRITING2:
        if (typewrite_next_char() && li == 7) {
            state = WAITING_KEYPRESS;
        }
        break;
    case WAITING_KEYPRESS:
        break;
    }
}

void SplashScene::draw() const {
    pico_set_color_draw(PICO_COLOR_WHITE);
    
    for (size_t lj = 0; lj <= (size_t)li; lj++) {
        std::string str = lj == (size_t)li ? LINES[li].substr(0, ci) : LINES[lj];
        if (str.empty()) continue;
        
        Pico_Rel_Rect txt_r = { '!', {(float)OFFS[lj].x, (float)OFFS[lj].y, 0, 8}, PICO_ANCHOR_NW, NULL };
        pico_output_draw_text(str.c_str(), &txt_r);
    }

    if (state == TYPEWRITING2 || state == WAITING_KEYPRESS) {
        // Ensure parent layers exist
        pico_layer_image_mode('=', "ufo", IMG_UFO);
        pico_layer_image_mode('=', "inv1", IMG_INV1);
        pico_layer_image_mode('=', "inv2", IMG_INV2);
        pico_layer_image_mode('=', "inv3", IMG_INV3);

        auto draw_sprite = [](const char* key_par, const char* name, float x, float y, float sw, float sh) {
            char key[256];
            snprintf(key, sizeof(key), "/crop/%s", name);
            Pico_Rel_Rect crop = { '!', {0, 0, sw, sh}, PICO_ANCHOR_NW, NULL };
            pico_layer_sub_mode('=', key, key_par, &crop);
            Pico_Rel_Rect dst = { '!', {x, y, sw, sh}, PICO_ANCHOR_C, NULL };
            pico_output_draw_layer(key, &dst);
        };

        draw_sprite("ufo",  "ufo",  71, 140, 24, 8);
        draw_sprite("inv1", "inv1", 71, 156, 8, 8);
        draw_sprite("inv2", "inv2", 71, 172, 11, 8);
        draw_sprite("inv3", "inv3", 71, 188, 12, 8);
    }
}

File: lua/scenes/Splash.lua
local class = require "class"
local assets = require "assets"

local SplashScene = class.new()
local splash_lines = {
    "PLAY",
    "SPACE INVADERS",
    "*SCORE ADVANCES TABLE*",
    "=? MYSTERY",
    "=30 POINTS",
    "=20 POINTS",
    "=10 POINTS",
    ""
}
local splash_offs = {
    {96, 64}, {56,  88}, {24, 120},  {80, 136},
    {80, 152}, {80, 168}, {80, 184}, { 0,  0}
}

function SplashScene:init(game_state)
    self.game_state = game_state
    self.state = "WAITING1"
    self.timer = 0
    self.li = 1
    self.ci = 0
end

function SplashScene:typewrite_next_char()
    if self.timer < 0.05 then return false end
    
    local line_len = string.len(splash_lines[self.li])
    self.ci = self.ci + 1
    if self.ci > line_len then
        self.li = self.li + 1
        self.ci = 0
    end
    self.timer = self.timer - 0.05
    return true
end

function SplashScene:process_event(event)
    if event.tag ~= 'key.dn' then return end

    if self.state == "WAITING_KEYPRESS" then
        local PlayScene = require "scenes.Play"
        self.game_state.score = 0
        self.game_state.set_scene(PlayScene:new(self.game_state))
    else
        self.state = "WAITING_KEYPRESS"
        self.li = 8
        self.ci = 0
    end
end

function SplashScene:update(delta)
    self.timer = self.timer + delta

    if self.state == "WAITING1" then
        if self.timer >= 1.5 then
            self.state = "TYPEWRITING1"
            self.timer = self.timer - 1.5
        end
    elseif self.state == "TYPEWRITING1" then
        if self:typewrite_next_char() and self.li == 3 then
            self.state = "WAITING2"
        end
    elseif self.state == "WAITING2" then
        if self.timer >= 1.5 then
            self.state = "TYPEWRITING2"
            self.timer = self.timer - 1.5
        end
    elseif self.state == "TYPEWRITING2" then
        if self:typewrite_next_char() and self.li == 8 then
            self.state = "WAITING_KEYPRESS"
        end
    end
end

function SplashScene:draw()
    pico.set.color.draw("white")
    
    for lj = 1, self.li do
        local str
        if lj == self.li then
            str = string.sub(splash_lines[lj], 1, self.ci)
        else
            str = splash_lines[lj]
        end
        if str ~= "" then
            local txt_r = { '!', x=splash_offs[lj][1], y=splash_offs[lj][2], w=0, h=8, anchor='NW' }
            pico.output.draw.text(str, txt_r)
        end
    end

    if self.state == "TYPEWRITING2" or self.state == "WAITING_KEYPRESS" then
        pico.layer.image('=', "ufo", assets.IMG_UFO)
        pico.layer.image('=', "inv1", assets.IMG_INV1)
        pico.layer.image('=', "inv2", assets.IMG_INV2)
        pico.layer.image('=', "inv3", assets.IMG_INV3)

        local function draw_sprite(key_par, name, x, y, sw, sh)
            local key = "/crop/" .. name
            local crop = { '!', x=0, y=0, w=sw, h=sh, anchor='NW' }
            pico.layer.sub('=', key, key_par, crop)
            local dst = { '!', x=x, y=y, w=sw, h=sh, anchor='C' }
            pico.output.draw.layer(key, dst)
        end

        draw_sprite("ufo",  "ufo",  71, 140, 24, 8)
        draw_sprite("inv1", "inv1", 71, 156, 8, 8)
        draw_sprite("inv2", "inv2", 71, 172, 11, 8)
        draw_sprite("inv3", "inv3", 71, 188, 12, 8)
    end
end

return SplashScene
